opal/corelib/runtime.js

```diff
@@ -601,13 +608,15 @@
         $set_proto(self, klass.$$prototype);
         return self;
       }
+    } else if ($truthy(name)) {
+      fun_name = name == "NilClass" ? "nil" : name
+      klass = Function("return function "+fun_name+" (){}")();
     } else {
       klass = function(){};
     }
-
-    if (name && name !== nil) {
-      $prop(klass, 'displayName', '::'+name);
-    }
+  
+    $prop_gs(klass, "name", false, $mod_name_getter);
+    $prop_gs(klass, "displayName", false, $mod_name_getter);
```

Comment:
Would be possible to define these on `Class` itself (or, even better, `Module`!) and rely on `$$name` dynamically?

`$set_proto(klass, Opal.Class.prototype);`

Reply:
No. `name` is set during function definition, which overrides setting on prototype.